[home page] Update Learn cards

ashleemboyer opened the pull request:
## 🔗 Relevant links

<!--
Include links to the branch preview, Asana task, and Figma designs wherever possible to make reviewing your PR easier.
When including the preview link, make sure to remove any '.' characters from the branch name:
  - Example: ks.my-branch -> ksmy-branch
-->

- [Preview link](https://dev-portal-git-ambupdate-certification-cards-hashicorp.vercel.app/) 🔎
- [Asana task](https://app.asana.com/0/1202114367927919/1202439144025013/f) 🎟️

## 🗒️ What

<!--
Briefly list out the changes proposed in this PR.
-->

- Updates the Learn cards on the home page to:
  - have their data loaded server-side via a slug
  - render the card based on the type of content (tutorial or collection)

## 🤷 Why

<!--
Describe why the changes proposed are needed. Some examples: new feature requested, refactor to make things easier later, styling tweaks requested by design, etc.
-->

- The data was previously hard-coded
- Tutorial cards were being shown for Collection content

## 🛠️ How

<!--
Dive into the approach you took, list resources you referenced, detail other approaches you tried but didn't end up going with, etc.
-->

- Added a `server.ts` file for `HomePageView` that loads featured learn data in the same way as the install views do
- Updated `HomePageView` to accept a new `featuredLearnCards` prop that's generated server side
- Updated `src/views/homepage/components/learn-section/index.tsx` to render the correct kind of card based on the content

## 🧪 Testing

<!--
Create a checklist for going through how to test your proposed changes. If there is anything to configure before interacting with the project in a browser, such as toggling feature flags, changing machine settings, or simulating behavior in browser dev tools, list those steps first.

- [ ] Step 1
- [ ] Step 2
- [ ] Step 3
- [ ] ...
-->

- [ ] Go to the home page
- [ ] Make sure there are now collection cards rendering in the learn section

## 💭 Anything else?

<!--
If there is anything you came across that you chose not to address in this PR but plan to soon, list those items here and any Asana tasks you created to go with them.
-->

It would be really nice to abstract this functionality into `lib` helpers instead of sourcing them from the install view and learn card helpers. Soon! 🤩

zchsh commented on src/views/homepage/components/learn-section/index.tsx:
```diff
@@ -51,17 +52,16 @@ export default function LearnSection({
-              >
-                <TutorialCard {...tutorial} />
+              <li key={id} className={s.tutorialsListItem}>
+                {type === 'collection' && <CollectionCard {...cardProps} />}
+                {type === 'tutorial' && <TutorialCard {...cardProps} />}
```
⛏️ Nitpick: This feels a little odd to me since it feels technically possible that nothing would render here. One approach could be to do like `const CardComponent = type === 'collection' ? CollectionCard : TutorialCard` in the `map` body, and then render `<CardComponent {...cardProps} />`.

For context: I think we had [a Slack thread](https://hashicorp.slack.com/archives/C02T80JTNS2/p1649872956699629), where it felt like we had alignment on wanting to avoid `&&` for conditional rendering (here's [an article I like on the pitfalls of &&](https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx))... though I think we still haven't set anything in stone, so definitely feel free to ignore this if it doesn't feel productive 👍

zchsh reviewed (dismissed):
🙌 ✅  This looks great to me!

Had some _very_ nitpick-y things, so definitely feel free to ignore 👍 🚀

kendallstrautman commented on src/pages/index.tsx:
```diff
@@ -1,5 +1,20 @@
+        collectionSlug: 'vault/associate-cert',
+      },
+      {
+        collectionSlug: 'vault/ops-pro-cert',
```
We should probably adjust all the data in this view to be editable via a json file, to follow similar patterns. I'm just thinking of authors needing to come and update content in this file, whereas an abstracted data file would be ideal. Of course, this is a long term thought, [not necessary for this PR. ](https://app.asana.com/0/1202276096167515/1202445483168876)

kendallstrautman commented on src/pages/index.tsx:
```diff
@@ -1,5 +1,20 @@
+
+export const getStaticProps: GetStaticProps = async () => {
+  return generateStaticProps({
+    featuredLearnContent: [
```
I'd agree for hedging against flexibility here. We could have something like an inline collections array of slugs. But also, this works for now.

kendallstrautman commented on src/views/homepage/server.ts:
```diff
@@ -0,0 +1,50 @@
+  )
+
+  // Transform featured tutorial and collection entries into card data
+  const featuredLearnCards: FeaturedLearnCard[] = featuredLearnContent.map(
```
If we keep this here we could note the duplication of this code in the install view, just as a 🔖 to be aware of the spread.

kendallstrautman reviewed (approved):
My vote would be to remove the tutorial card option from this implementation and just do collection cards. But approving as to not block this work 👌

braydenlove reviewed (approved):
Nice job, approved.

ashleemboyer commented:
@kendallstrautman @zchsh @braydenlove I've updated this PR to now only show/allow collection cards! I've also added a `content.json` file as a head start to working on this new task: [Refactor homepage data to json file](https://app.asana.com/0/1202114367927919/1202445483168876/f). This is ready for re-review ✨

kendallstrautman commented on src/views/homepage/components/learn-section/index.tsx:
```diff
@@ -51,17 +52,15 @@ export default function LearnSection({
           style={
             {
-              '--tutorials-count': tutorials.length,
+              '--tutorials-count': collectionCards.length,
```
⛏️ Nit: Perhaps we should update this css var? `--collection-cards-count`?  (non blocking)

ashleemboyer replied:
Good point! Glazed right over this one 😆

kendallstrautman commented on src/views/homepage/server.ts:
```diff
@@ -0,0 +1,41 @@
+
+  // Transform collection entries into card data with `collectionSlugs` order
+  const collectionCards = collectionSlugs.map((collectionSlug: string) => {
+    const thisCollection = collections.find(
```
Might be worth a comment here re: sorting 🤷‍♀️

kendallstrautman reviewed (approved):
Looks great! Ty for reworking this 🙇

zchsh reviewed (approved):
🙌  +1 on the extra optional push here for making some tough calls and doing a super clean rework! 💯

braydenlove reviewed (approved):
_`APPROVED`_ ✅
